### PR TITLE
Plugin docs (for plugin developers)

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Just wanted to get on top of writing some plugin docs.  I believe @jni was planning on writing the docs for napari contributors and developers that would explain how the napari plugin architecture is implemented.  I created a stub file here for him to do that.  The file that I created is a "guide on creating plugins" aimed at potential plugin developers.

I'll leave this as WIP until at least #937 merges, in case anything changes about the hookspecs.

## Type of change

- [x] docs

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
Maybe this is premature, but when we have support for plugins that take in an image layer & return a thing, I'd prefer that as the example featured in the developer docs. Long term, it's probably close to what most plugin developers want to do. Does that sound reasonable?

**Comment by jni** [comment]
@GenevieveBuckley It *is* premature, I think! :stuck_out_tongue: I think the documentation will be split up and there will be a separate document for that functionality. I think both kinds of plugins are important, and interactive plugins as well — we should have examples for all major use cases.

**Comment by tlambert03** [comment]
Yeah, we should have both examples somewhere eventually ... but it may be a while until we have our first concept of a plugin that takes a layer and returns a thing

**Comment by GenevieveBuckley** [comment]
> 
> Yeah, we should have both examples somewhere eventually ... but it may be a while until we have our first concept of a plugin that takes a layer and returns a thing

Oh, for sure

**Comment by tlambert03** [comment]
This is ready for review.  I updated the `hook_specification` docstring and the main docs to account for the recent change to the `napari_get_reader` hook specification (now accepts both a `str` and `List[str]`.  To see the full formatting, you'll need to check out this PR, run `make docs`, then `open docs/build/html/index.html ` but you can also of course just read the file in this PR

**Comment by kevinyamauchi** [comment]
Hey, @tlambert03! This looks really nice. I am going to try to make a plugin over the next couple of days (load deeplabcut data into a `Points` layer). I can review as I use your docs to make the plugin :)

I am super excited about all of this great work you have done on the plugins!

**Comment by tlambert03** [comment]
with most of the plugin PRs merged, I took another look at this and made sure everything is up to date and likely to remain stable.  Looks good to me, so feel free to review.  @jni, note the `for_napari_developers.rst` stub where you might eventually put some dev-focused docs as we discussed.

**Comment by sofroniewn** [comment]
Are there any edits needed to this now that #1023 / #1024 are in? If not I will start reviewing asap

**Comment by tlambert03** [comment]
> Are there any edits needed to this now that #1023 / #1024 are in? If not I will start reviewing asap

I re-read it this morning and I think it's up-to-date (as far as I can tell)

File: docs/source/conf.py
Comment (jni) [inline docs/source/conf.py]:
```suggestion
# add_module_names = False avoids showing the full path to a function or class
```

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]:
```suggestion
*hook*, then, is the place within napari where we
```

😉

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]:
```suggestion
well-documented function signatures that define the API (or
```

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst:42]: Love the above four points. =)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: Can you include this docstring here? Or is it too much? Maybe you can link the function here also. (I see the link above the code block, but I would expect "docstring" to be clickable here. =)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: maybe specify where this is explained?

Comment (jni) [reply docs/source/plugins/for_plugin_developers.rst]: And, seeing that the code requires explanation, I wonder whether we should show importing `napari_hook_implementation` here, which is more self-explanatory, even if only on a superficial level. Then, later where it is explained, you can say, "and, bonus, you don't actually need to import napari to create a plugin! tada!" and explain the fancier pluggy version. What do you think? (Also @sofroniewn)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: I'd remove this comment as well.

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: I would move this definition above the `napari_get_reader` call and rename it `ext_file_reader`. To quote @stefanv from long ago, it's valid for the Python parser to define the function under the place where it's called, but it breaks my brain's parser. =P (He told *me* when I'd done the exact same thing. 😂)

Comment (jni) [reply docs/source/plugins/for_plugin_developers.rst]: Also, I don't love that `convert_bytes_to_numpy` is undefined. Maybe just do a real plugin here for .npy files using np.load?

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [reply docs/source/plugins/for_plugin_developers.rst]: (save it for when the usage of that form is documented.)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: maybe link here to the document you showed us, I think from PyPA (?), when we were first discussing this?

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst:248]: I feel like this needs a follow-up line:
```suggestion
   pip install napari mypackage

which is 🎵 ah-maziiing! 🎵 
```

😀

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: You should decide whether you want $ prompts in front of bash code, or not (see above). I think I prefer not, for copy-paste convenience. Ironically, you'd never want to copy the bash code above this one (pip install napari mypackage), but you *would* want to copy the above. =)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst]: Is there a link for this format? Seems very mysterious otherwise... =)

Maybe also link to aicsimageio for a more elaborate example? =)

File: docs/source/plugins/for_plugin_developers.rst
Comment (jni) [inline docs/source/plugins/for_plugin_developers.rst:291]: Is that the right place? Or maybe image.sc?

File: napari/plugins/hook_specifications.py
Comment (jni) [inline napari/plugins/hook_specifications.py]:
```suggestion
hook implementations may accept *fewer* arguments than defined in the
```

https://www.youtube.com/watch?v=qmVnr7rsWrE

😁

**Review by jni — APPROVED** [review]
@tlambert03 I've made minor fixes and suggestions. Approving so you're not held up from merging when fixed. =)

btw, thank you, the guide is a *stellar* document. Can't wait to tweet it! =D

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: I'll make `docstring` a clickable link.  rather not copy the full thing here

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: will switch to a numpy reader plugin

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: sure.  I was a little conflicted... since I'd really rather people not default to depending on napari for that line alone.  But I agree it reads better... and until a plugin like this can be used by something other than napari, not depending on napari is of minor benefit.

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst:248]: lol.  serious??  I'll do it!

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: no.  I'd prefer to be inconsistent thank you. 😂 ... removed the `$`

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst:291]: image.sc def.  added.  but also suggest to open an issue with bug/feature requests.

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: added a link and also highlight aicsimageio.

**Review by tlambert03 — COMMENTED** [review]
Thanks as always for the thorough look.  It's much better as a result.  I made pretty much all the changes.  Only remaining questions are
- @sofroniewn, do you agree with switching to using the internal `napari.plugins.napari_hook_implementation` as the default example, then showing how you can also import directly from pluggy?
- @jni, did you really want me to put the playful music notes in?? 😄(cause i'll do it!)

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [reply docs/source/plugins/for_plugin_developers.rst:248]: umm ..... not so sure about this one ;-)

Comment (sofroniewn) [reply docs/source/plugins/for_plugin_developers.rst:248]: but it is amazing :-)

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [inline docs/source/plugins/for_plugin_developers.rst:38]: You mention "one of two methods" here, but don't say which they are, could you maybe say something and then have the link, otherwise I find myself wanting more information here

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [inline docs/source/plugins/for_plugin_developers.rst]:
```suggestion
contract that you can follow, and know that napari will handle the rest.
```

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [reply docs/source/plugins/for_plugin_developers.rst:38]: Maybe just mention a "naming convention" or "package metadata"

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [reply docs/source/plugins/for_plugin_developers.rst:248]: maybe just "which is amazing!"

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [inline docs/source/plugins/for_plugin_developers.rst:15]: Do you want to mention the existence of the cookiecutter up here with a link below?

File: docs/source/plugins/for_plugin_developers.rst
Comment (sofroniewn) [inline docs/source/plugins/for_plugin_developers.rst]:
```suggestion
community plugins developed for napari.  This plugin takes advantage of
```

I prefer community over third-party, curious what others think. Feels more inclusive

**Review by sofroniewn — APPROVED** [review]
Looks good to me, left some mostly minor comments!

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst]: Yeah, I started there actually... I don't know why but somehow it felt like it was "spreading credit" for Jackson's work lol... but I also prefer the word community

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst:15]: added

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst:38]: added

File: docs/source/plugins/for_plugin_developers.rst
Comment (tlambert03) [reply docs/source/plugins/for_plugin_developers.rst:248]: while I am equally enthused, I think I'll leave it out ;)

**Comment by jni** [comment]
> @jni, did you really want me to put the playful music notes in?? 😄(cause i'll do it!)

I would have been into it, but as long as you know it, that's good enough. =)

**Pull request merged**
